Read out real strides from compilation result, rather than real args

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104971
* __->__ #105010
* #105009

This prefigures a refactor that will move the backward compilation
to entirely ahead of time, so I need to extract these strides some
other way.  Straight from the compiler's mouth will do it.

I can't easily get the information via the return result of `fw_compiler` without changing the calling convention, so instead I smuggle it via TracingContext. TracingContext may be None when we are compiling patterns for the joint graph pattern matcher.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8

ezyang committed:
Read out real strides from compilation result, rather than real args

This prefigures a refactor that will move the backward compilation
to entirely ahead of time, so I need to extract these strides some
other way.  Straight from the compiler's mouth will do it.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

shunting314 commented on torch/_guards.py:
```diff
@@ -482,6 +482,11 @@ def __init__(self, fake_mode):
+        # compiler to aot_autograd
+        # Per output, what the compiler specified stride of the output is,
+        # or None if no stride is known
+        self.output_strides: Optional[List[Optional[List[int]]]] = None
```
Right now we do layout optimization only if dynamic shape is disabled. But we are working on resolving that (blocked on the split reduction support). So, should we set the type here to consider SymInt?

ezyang replied:
yeah this can have symints, will amend

shunting314 commented on torch/_inductor/compile_fx.py:
```diff
@@ -496,6 +496,15 @@ def fx_codegen_and_compile(
+                # Return the output strides to the caller via TracingContext
+                assert len(context.output_strides) == 0
+                for out in graph.graph_outputs:
+                    if hasattr(out, "layout"):
```
Do we see cases that `out` does not have a `layout` attribute? Does this happen for some SymInt returned?

ezyang replied:
yes, can have symint return. Maybe there is something better than hasattr to do here, but a lot of inductor code is written this way. Hard to be more precise without typing 👹

ezyang replied:
OK, turns out I can't, because strides in inductor are sympy.Symbol, not SymInt. So I am just going ahead and storing the hints here only. When you fix this to do permutations instead, probably can make this a little less fragile.

shunting314 replied:
We can not do permutations instead because of some non 'dense' activations? This happens for some real models.

ezyang replied:
There are a few ways we can do this. If we only ever change layout on dense tensors, we can make `output_strides` be None *unless* we changed the stride. Then the permutation is always defined.

shunting314 replied:
> If we only ever change layout on dense tensors, ...

I think this can not be guaranteed in inductor right now.  A non dense tensor's layout may get changes because of the layout change of its upstream tensors.  And inductor does not force eager stride for that non dense tensor because of code here: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/graph.py#L688 . I guess the algorithm will be tricky to restride a non-dense tensor.

ezyang replied:
OK will need to think this through carefully. I'll start a doc.

ezyang replied:
Upon further reflection, all of this is moot if we have compiled backwards, so let's not touch it unless it's causing someone problems.

shunting314 commented on torch/_guards.py:
```diff
@@ -520,6 +525,20 @@ def current_frame(frame_summary):
+        if tc is None:
+            yield None
+            return
+        old_output_strides = tc.output_strides
```
Just want to raise a concern about nested compiling where we may need a stack for output_strides. But I can not come up with a realistic example. So probably the current implementation is good enough

ezyang replied:
This implicitly is a stack via old_output_strides!

shunting314 replied:
ah, right, that's the whole point of context manager after all... haha

shunting314 commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3003,20 +3004,37 @@ def call_compiled_backward():
                             continue
 
-                        # Comparing ph_arg.stride() with real_arg.stride() directly may
+                        if forward_saved_for_backwards_strides is None:
```
Move this check out of the for loop may be slightly faster.

ezyang replied:
I was trying to save myself an indent 😂  This is only during compilation so I don't think it matters much

shunting314 commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3003,20 +3004,37 @@ def call_compiled_backward():
+                        if real_stride is None:
+                            continue
+
+                        assert _get_hints(real_stride) == all_args[i].stride(), f"{real_stride} {all_args[i].stride()}"
```
I assume you will remove this soon since the may point of the change is we won't have access to `all_args`?

ezyang replied:
Yup, this is just for CI here.

shunting314 commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3003,20 +3004,37 @@ def call_compiled_backward():
 
+                    forward_saved_for_backwards_strides = None
+                    if fwd_output_strides is not None:
+                        forward_saved_for_backwards_strides = fwd_output_strides[
+                            CompiledFunction.metadata.tensors_saved_for_backwards_slice
+                        ]
```
This looks nice!

ezyang committed:
Update on "Read out real strides from compilation result, rather than real args"


This prefigures a refactor that will move the backward compilation
to entirely ahead of time, so I need to extract these strides some
other way.  Straight from the compiler's mouth will do it.

I can't easily get the information via the return result of `fw_compiler` without changing the calling convention, so instead I smuggle it via TracingContext. TracingContext may be None when we are compiling patterns for the joint graph pattern matcher.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

ezyang committed:
Update on "Read out real strides from compilation result, rather than real args"


This prefigures a refactor that will move the backward compilation
to entirely ahead of time, so I need to extract these strides some
other way.  Straight from the compiler's mouth will do it.

I can't easily get the information via the return result of `fw_compiler` without changing the calling convention, so instead I smuggle it via TracingContext. TracingContext may be None when we are compiling patterns for the joint graph pattern matcher.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
